cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -1527,4 +1631,31 @@ private void emitTimeMetrics(String namespace, String appName, String metricName
     long timeTaken = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime);
     metricsCollectionService.getContext(tags).gauge(metricName, timeTaken);
   }
+
+  public List<ApplicationId> markAppsAsLatest(List<ApplicationId> appIds) throws Exception {
+    ExecutorService executorService = Executors.newFixedThreadPool(AppFabric.DEFAULT_WORKER_THREADS);
+    List<ApplicationId> failedAppIds = new ArrayList<>();
+
+    for (ApplicationId appId: appIds) {
+      executorService.execute(() -> {
```

Comment:
the store operation is a fast one on a single row, should not be doing those in separate threads. Additionally, it is important that all of the marking happens atomically in one transaction. The store should have a method that takes the list of application ids.

Reply:
Also keep in mind that a `List<ApplicationId>` means somebody could potentially pass in multiple versions of the same application, which is invalid. We will either need to check for that in the logic, or use some other class which makes that impossible.

Reply:
Added validation to check for duplicates.